<a href="https://colab.research.google.com/github/dSudario9/Modelos-Supervisionados-Diversos/blob/main/Rede_Neural_MLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***Criando Modelo de Classificação MLP Rede Neural ***

In [ ]:
# import da Bibliotecas 

import pandas as pd 

EDA

In [ ]:
# Carregando base de dados 
df = pd.read_excel("/content/Bank_Personal_Loan_Modelling.xlsx",sheet_name="Data")

In [ ]:
df.head()

Age  Experience  Income  ...  CD Account  Online  CreditCard
9    34           9     180  ...           0       0           0
16   38          14     130  ...           0       0           0
18   46          21     193  ...           0       0           0
29   38          13     119  ...           1       1           1
38   42          18     141  ...           1       1           0

[5 rows x 12 columns]

In [ ]:
# Excluindo variaveis nao uteís 

df= df.drop(columns=["ID", "ZIP Code"])


In [ ]:
# Visualização de estatistica basica. 
df.describe()

Age   Experience  ...       Online   CreditCard
count  5000.000000  5000.000000  ...  5000.000000  5000.000000
mean     45.338400    20.104600  ...     0.596800     0.294000
std      11.463166    11.467954  ...     0.490589     0.455637
min      23.000000    -3.000000  ...     0.000000     0.000000
25%      35.000000    10.000000  ...     0.000000     0.000000
50%      45.000000    20.000000  ...     1.000000     0.000000
75%      55.000000    30.000000  ...     1.000000     1.000000
max      67.000000    43.000000  ...     1.000000     1.000000

[8 rows x 12 columns]

In [ ]:
# Distribuição das Variaveis Alvo ou dados de entrada e saida.

df ["Personal Loan"].value_counts()

# Observando que a base esta desbalanceada .

0    4520
1     480
Name: Personal Loan, dtype: int64

In [ ]:
df.info('Personal Loan')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Age                 5000 non-null   int64  
 1   Experience          5000 non-null   int64  
 2   Income              5000 non-null   int64  
 3   Family              5000 non-null   int64  
 4   CCAvg               5000 non-null   float64
 5   Education           5000 non-null   int64  
 6   Mortgage            5000 non-null   int64  
 7   Personal Loan       5000 non-null   int64  
 8   Securities Account  5000 non-null   int64  
 9   CD Account          5000 non-null   int64  
 10  Online              5000 non-null   int64  
 11  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(11)
memory usage: 468.9 KB


In [ ]:
# Pelo fato da base ser desbalanceada, não iremos usar toda a classe 0.
# Iremos pegar uma amostra e juntar com o registros da classe 1 para podermos modelar de forma mais justa.

df_classe_1 = df[df["Personal Loan"] == 1]
df_classe_0 = df[df["Personal Loan"] == 0].sample(n=1500)
df = df_classe_1.append(df_classe_0)

Criando Modelo 


In [ ]:
# Dividindo preditores (X) e variável target (Y) dados de entrada e dados de saida 

X = df.drop(columns=["Personal Loan"])
Y = df["Personal Loan"]

Separando bade de Train e Test 

In [ ]:
# Dividindo a base em treino e teste

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=7)

print(f"Shape X_train: {X_train.shape}")
print(f"Shape y_train: {y_train.shape}")
print(f"Shape X_test: {X_test.shape}")
print(f"Shape y_test: {y_test.shape}")

Shape X_train: (1584, 11)
Shape y_train: (1584,)
Shape X_test: (396, 11)
Shape y_test: (396,)


In [ ]:
# Observando que na base de dados a varial Personal Loan precisa de normalização uma ver que as casas deciamais estão diferentes. 

# Normalização das variáveis com StandardScaler

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Transformanda para dataframe para visualização
X_train = pd.DataFrame(X_train_scaled,columns = X_train.columns)
X_test = pd.DataFrame(X_test_scaled,columns = X_test.columns)

Criando Classificador 

In [ ]:
# Classificador Multi Layer Perceptron

from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

clf = MLPClassifier(hidden_layer_sizes=(300, 100, 100),activation='relu')


Treinando Modelo 

In [ ]:
clf.fit(X_train,y_train)

MLPClassifier(hidden_layer_sizes=(300, 100, 100))

Realizando predição 

In [ ]:
y_pred = clf.predict(X_test)

Acurácia

In [ ]:
accuracy_score(y_test ,y_pred)*100

96.71717171717171

Como observamos neste modelo conseguimos um resultado satisfatorio com acurácia de 96.71717171717171%. alteramos os  tamanhos de camadas escondidas de 100,50,20  para 300, 100,100, foi o melhor resultado que encontramos com o parametro .

In [ ]:
# Através da análise das variáveis e do dicionário, algums informações relevantes:


# Variáveis categóricas:
# Personal Loan (Empréstimo Pessoal) - Este cliente aceitou o empréstimo pessoal oferecido na última campanha? Esta é a nossa variável alvo
# Securities Account (Conta de títulos) - O cliente possui conta de títulos no banco?
# CD Account (Conta de CD) - O cliente tem uma conta de certificado de depósito (CD) no banco?
# Online - O cliente usa serviços de banco pela Internet?
# Credit Card - (Cartão de crédito) - O cliente usa um cartão de crédito emitido pelo UniversalBank?

# Variáveis contínuas:

# Age (Idade) Idade do cliente
# Experience (Experiência) - anos de experiência
# Income (Renda) - Renda anual em dólares
# CCAvg (CCAvg) - gasto médio com cartão de crédito
# Mortage (Hipoteca) - Valor da hipoteca da casa

# Variáveis ​​categóricas ordinais:
# Family (Família) - tamanho da família do cliente
# Education (Educação) - nível de educação do cliente


# A variável ID não adiciona nenhuma informação interessante. 
# Não há associação entre o ID de cliente de uma pessoa e o empréstimo, também não fornece nenhuma conclusão geral para futuros clientes de empréstimo em potencial. 
# Podemos desconsiderar essas informações para o modelo.
# Zip code também pode ser desconsiderada, uma vez que a informação de distnância entre zip code pode não ser realidade